### ISSUE TITLE
Race condition between creation of ActorSystem and  Cluster.clusterEvents

### ISSUE THREAD

**Issue by politrons** [zio/zio-akka-cluster#6]
Hi again,

I think I´m facing a possible Race condition using
```

  val clusterEvents: ZIO[ActorSystem, Throwable, Queue[ClusterDomainEvent]] =
    Queue.unbounded[ClusterDomainEvent].tap(clusterEventsWith)
```

This queue is filled with the events of actions over the cluster.

The problem is that the `ActorSystem` has been created previously passed as provider into the Monad `ZIO[ActorSystem, Throwable, Any]` 

`mainProgram.unsafeRunToFuture(bootableProgram.provide(actorSystem))
`

so it happening sometimes that the Member is added into the cluster before I subscribe into the Queue so I´m missing that event.

```
events <- Cluster.clusterEvents
event <-events.take

```
Any suggestion to prevent that RC?

I guess not pass the `ActorSystem `as Environment when I run the monad and pass into the `Cluster.clusterEvents,` is an option but definitely less cool option :(

```
 events <- Cluster.clusterEvents.provide(ActorSystem.create("bla"))
event <-events.take
```

Regards.

**Comment by ghostdogpr** [comment]
There is actually a parameter that lets you get the events that happened in the past to reach the current state: https://doc.akka.io/docs/akka/current/cluster-usage.html (look for `InitialStateAsEvents`).

Currently I’m not exposing this option but I could add it. What do you think?

Another workaround is to call `clusterState` after you get the queue but before `take` to check if you’re already in the state you want.

**Comment by politrons** [comment]
I fixed, tough not that cool solution as you suggested before with Events, but still working, doing polling.

```
private def waitUntilClusterIsFormed: ZIO[ActorSystem, Throwable, Unit.type] = {
    for {
      actorSystem <- ZIO.environment[ActorSystem]
      _ <- ZIO.effect {
        logger.info(null, s"ZIO Akka Cluster: Waiting for cluster to be formed")
        var allMembersUp = false
        while (!allMembersUp) {
          val state = mainProgram.unsafeRun(Cluster.clusterState.provide(actorSystem))
          if (state.members.size >= getActorSystemProperty(actorSystem, "akka.cluster.min-nr-of-members").toInt) {
            logger.info(null, s"ZIO Akka Cluster: Minimum number to create Cluster reach.")
            clusterMembers = clusterMembers ++ state.members.toList
            allMembersUp = true
          }
          Thread.sleep(100)
        }
      }
    } yield Unit
  }
```

If you think in a better solution let me know

**Comment by politrons** [comment]
> InitialStateAsEvents

I believe it would be a great idea, I would not mind even help you out (in a week I´m on Holidays 😄  )

But indeed have the capacity to get all previous events it could prevent a lot of pain in this sort of RC.

Meanwhile  I will try your suggestion, although the one that I told you before works.

**Comment by ghostdogpr** [comment]
@politrons I created #10 for it

**Comment by ghostdogpr** [comment]
@politrons released 0.1.1 with this change
